File: src/main/java/dealing/with/generalization/Event.java
package dealing.with.generalization;

public abstract class Event {

    protected final String description;
    private boolean running;
    private long startTime = 0;
    private long stopTime = 0;

    protected Event(String description) {
        this.description = description;
    }

    public void start() {
        if (running) {
            throw new IllegalStateException("Concert is already in progress");
        }
        this.startTime = System.currentTimeMillis();
        running = true;
    }

    public void stop() {
        if (running) {
            running = false;
            stopTime = System.currentTimeMillis();
        }
    }

    public long getElapsedTime() {
        return running ? (System.currentTimeMillis() - startTime) : (stopTime - startTime);
    }
}

File: src/scala/src/main/scala/dealing/with/generalization/Event.scala
package org.ythirion.refactoring.journey
package dealing.`with`.generalization

import scala.util.{Failure, Success, Try}

abstract class Event {
  private var running: Boolean = false
  private var startTime: Long = 0
  private var stopTime: Long = 0

  def start(): Try[Boolean] = {
    if (running) {
      Failure(new IllegalStateException(s"${this.getClass.getSimpleName} is already in progress"))
    }
    else {
      startTime = System.currentTimeMillis()
      running = true

      Success(running)
    }
  }

  def stop(): Unit = {
    if (running) {
      running = false
      stopTime = System.currentTimeMillis()
    }
  }

  def calculateElapsedTime: Long = {
    if (running) System.currentTimeMillis() - startTime
    else stopTime - startTime
  }
}